[OID4VCI]: Realm-Configurable Time-Claim Normalization (Randomize/Round) to Mitigate Correlation

Awambeng opened the pull request:
This PR implements realm-configurable correlation mitigation for time-related claims (iat, nbf, exp).

Key Changes:

- Adds TimeClaimNormalizer supporting rounding (minute/hour/day) or randomization within a configurable window.
- Applied at issuance in the issuer endpoint and leverages existing SD-JWT mappers.
- Includes unit and integration tests for both formats.
- Updates OID4VC documentation with changes

Closes #43399

Awambeng committed:
feat(oid4vc): realm-configurable time-claim normalization (round/randomize) and tests

Add TimeClaimNormalizer with strategies: off, randomize(window), round(unit).
Apply normalization to issuanceDate in issuer endpoint and to exp in JWT-VC builder.
Add OID4VCTimeClaimsNormalizationMapper for SD-JWT claims and JWT-VC exp.
Fix OID4VCIssuedAtTimeClaimMapper truncation logic and supported units.
Add unit and integration tests validating rounding for JWT-VC and SD-JWT.

Closes #43399

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Address @IngridPuppet review comments

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Address review comments

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Address review comments

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Address review comments

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

ensure KeycloakSession is not null when defaulting issuance time

Added a check to throw an IllegalStateException if the KeycloakSession is null during issuance time normalization in JwtCredentialBuilder. Updated tests to reflect this requirement and ensure proper handling of issuanceDate.

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

refactor(oid4vc): enhance JwtCredentialBuilder to support custom issuance time normalization

Updated JwtCredentialBuilder to accept a UnaryOperator for issuance time normalization, allowing for more flexible handling of issuance dates. Removed the requirement for KeycloakSession in the default constructor, improving usability. Adjusted tests to validate the new behavior, ensuring that the normalized issuance time is correctly applied when no issuance date is provided.

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Awambeng committed:
Address @Captain-P-Goldfish review comments

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Awambeng commented:
Hi @Captain-P-Goldfish, I’ve addressed the comments you left. Could you please take another look?

@mposolda, hello! Could you please check the CI? It seems there’s a broken link in the docs, but it doesn’t appear to be related to my changes.

Awambeng committed:
chore: fix formatting violations with spotless

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

mposolda commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/TimeClaimNormalizer.java:
```diff
@@ -0,0 +1,156 @@
+
+    private static final Logger logger = Logger.getLogger(TimeClaimNormalizer.class);
+
+    public static final String ATTR_STRATEGY = "oid4vci.time.claims.strategy";
```
Is it possible to move those constants to the `Oid4VciConstants` to ideally have single place with the various realm/client attributes related to OID4VCI?

mposolda reviewed (commented):
@Awambeng Thanks for the PR!

I've added one review comment inline. Can you please check it?

Besides that, I assume that this will need follow-up PR to update also UI with the new attributes? Can you please create follow-up GH issue for that?

Awambeng committed:
Address @mposolda review comment

Signed-off-by: Awambeng <awambengrodrick@gmail.com>

Awambeng commented:
> @Awambeng Thanks for the PR!
> 
> I've added one review comment inline. Can you please check it?
> 
> Besides that, I assume that this will need follow-up PR to update also UI with the new attributes? Can you please create follow-up GH issue for that?

Hi @mposolda, thank you very much for the review!

I’ve addressed the comments you left, but unfortunately I’m still encountering the CI failure mentioned above. Could you please take another look?

Regarding the integration of the additional attributes in this PR: on the UI side, we have this ticket https://github.com/keycloak/keycloak/issues/43900, which covers the attributes included here. It’s currently being reviewed internally, and we expect to submit a PR before the end of this week.

tnorimat commented on services/src/test/java/org/keycloak/protocol/oid4vc/issuance/TimeClaimNormalizerTest.java:
```diff
@@ -0,0 +1,115 @@
+        assertFalse("Normalized time should not be after original time", normalized.isAfter(orig));
+        // should be within the randomization window
+        long diffSeconds = orig.getEpochSecond() - normalized.getEpochSecond();
+        assertFalse("Randomized time should be within window", diffSeconds > 3600);
```
In my opinion, the following might be good for the readability. But it is up to you to do that or not.
```
        assertTrue("Randomized time should be within window", diffSeconds <= 3600);
```

Awambeng committed:
Address review comments

Signed-off-by: Awambeng <awambengrodrick@gmail.com>

Captain-P-Goldfish commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCIssuedAtTimeClaimMapper.java:
```diff
@@ -130,8 +130,8 @@ public void setClaimsForCredential(VerifiableCredential verifiableCredential,
-                .flatMap(config -> Optional.ofNullable(config.get(TRUNCATE_TO_TIME_UNIT_KEY)))
-                .filter(val -> !val.isEmpty())
+                .map(config -> config.get(TRUNCATE_TO_TIME_UNIT_KEY))
+                .filter(val -> val != null && !val.isEmpty())
```
`val != null` is not necessary it can simply be `filter(val -> !val.isEmpty())`

Awambeng committed:
Updated the truncateToTimeUnit configuration to include SECONDS as a supported option. Adjusted the filtering logic for the time unit configuration to ensure it handles empty values correctly.

Signed-off-by: Awambeng <awambengrodrick@gmail.com>

Awambeng commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCIssuedAtTimeClaimMapper.java:
```diff
@@ -75,7 +75,7 @@ public class OID4VCIssuedAtTimeClaimMapper extends OID4VCMapper {
-        truncateToTimeUnit.setHelpText("Truncate time to the start of the selected unit. Supported: MINUTES, HOURS, HALF_DAYS, DAYS, WEEKS, MONTHS, YEARS. Such as to prevent correlation of credentials based on this time value.");
+        truncateToTimeUnit.setHelpText("Truncate time to the start of the selected unit. Supported: SECONDS, MINUTES, HOURS, HALF_DAYS, DAYS, WEEKS, MONTHS, YEARS. Such as to prevent correlation of credentials based on this time value.");
         truncateToTimeUnit.setType(ProviderConfigProperty.LIST_TYPE);
         truncateToTimeUnit.setOptions(List.of("MINUTES", "HOURS", "HALF_DAYS", "DAYS", "WEEKS", "MONTHS", "YEARS"));
```
> `SECONDS` still missing in the options-list

Fixed.

Captain-P-Goldfish reviewed (approved):
LGTM

tnorimat reviewed (commented):
@Awambeng Hello, I added some minor comments. Could you check them?

Awambeng commented on docs/documentation/server_admin/topics/oid4vci/vc-issuer-configuration.adoc:
```diff
@@ -166,6 +166,61 @@ curl -X PUT "https://localhost:8443/admin/realms/oid4vc-vci" \
+
+| `oid4vci.time.randomize.window.seconds`
+| `86400`
+| When strategy is `randomize`, subtract a random number of seconds (0 to N) from the original timestamp to mitigate correlation attacks.
```
Yes, `N` represents the randomized window. I’ve applied the change as suggested. Thank you!

Awambeng commented on services/src/test/java/org/keycloak/protocol/oid4vc/issuance/TimeClaimNormalizerTest.java:
```diff
@@ -0,0 +1,115 @@
+        Instant orig = Instant.parse("2025-01-02T22:00:00Z");
+        TimeClaimNormalizer n = new TimeClaimNormalizer(TimeClaimNormalizer.Strategy.RANDOMIZE, 3600L, TimeClaimNormalizer.RoundUnit.DAY);
+        Instant normalized = n.normalize(orig);
+        assertFalse("Normalized time should not be after original time", normalized.isAfter(orig));
```
Same explanation as mentioned here: https://github.com/keycloak/keycloak/pull/43834#discussion_r2544615480

Awambeng commented on services/src/test/java/org/keycloak/protocol/oid4vc/issuance/TimeClaimNormalizerTest.java:
```diff
@@ -0,0 +1,115 @@
+
+        // we always randomize within the window (0 to 3600 seconds back)
+        // So normalized should be <= original and within the window
+        assertFalse("Normalized time should not be after original time", normalized.isAfter(orig));
```
Same explanation as mentioned here: https://github.com/keycloak/keycloak/pull/43834#discussion_r2544615480

tnorimat replied:
Thank you for the clarification.

Awambeng commented on services/src/test/java/org/keycloak/protocol/oid4vc/issuance/TimeClaimNormalizerTest.java:
```diff
@@ -0,0 +1,115 @@
+        // So normalized should be <= original and within the window
+        assertFalse("Normalized time should not be after original time", normalized.isAfter(orig));
+        long diffSeconds = orig.getEpochSecond() - normalized.getEpochSecond();
+        assertFalse("Randomized time should be within window", diffSeconds > 3600);
```
done

Awambeng committed:
Address @tnorimat review comments

Signed-off-by: Awambeng <awambengrodrick@gmail.com>

tnorimat reviewed (approved):
LGTM

mposolda reviewed (approved):
@Awambeng @Captain-P-Goldfish @tnorimat Thanks for this PR and for all the reviews!
